### PR TITLE
[p2p] Cache compact filters

### PR DESCRIPTION
**Pull request by cloudhead**

### CODE REVIEW

File: p2p/src/protocol/cbfmgr.rs [hunk 1]
```diff
@@ -225,12 +228,15 @@ pub enum GetFiltersError {
...
     pub request_timeout: Timeout,
+    /// Filter cache size, in bytes.
+    pub filter_cache_size: usize,
 }
 
 impl Default for Config {
     fn default() -> Self {
         Self {
             request_timeout: Timeout::from_secs(30),
+            filter_cache_size: 1024 * 1024, // 1 MB.
```
Comment (vincenzopalazzo) [inline p2p/src/protocol/cbfmgr.rs:239]: Do you think that this size it is better put in the protocol configuration and set 1MB as the default value?

File: p2p/src/protocol.rs [hunk 2]
```diff
@@ -1098,7 +1098,11 @@ impl<T: BlockTree, F: Filters, P: peer::Store> traits::Protocol for Protocol<T,
                 }
                 Command::Rescan { from, to, watch } => {
                     debug!(target: self.target, "Received command: Rescan({:?}, {:?})", from, to);
-                    self.cbfmgr.rescan(from, to, watch, &self.tree);
+
+                    // A rescan with a new watch list may return matches on cached filters.
+                    for hash in self.cbfmgr.rescan(from, to, watch, &self.tree) {
```
Comment (vincenzopalazzo) [inline p2p/src/protocol.rs:1103]: mh sorry for the question here, but this looks like magic for me. How the following call can work in a if it is a void function?

`self.cbfmgr.rescan(from, to, watch, &self.tree)`

P.S: Stupid question I know!

**Review by vincenzopalazzo — APPROVED** [review]
LGTM, I have only a couple of stupid questions.

ack https://github.com/cloudhead/nakamoto/pull/53/commits/6834070f8f301db23ab0229318c6b60f3ff7cdae

File: p2p/src/protocol/cbfmgr.rs (hunk 1 above)
Comment (cloudhead) [reply p2p/src/protocol/cbfmgr.rs:239]: That's the plan, this parameter will be passed down from the `protocol::Config` into the `cbfmgr::Config`

File: p2p/src/protocol.rs (hunk 2 above)
Comment (cloudhead) [reply p2p/src/protocol.rs:1103]: It is no longer void with this changeset, since `rescan` immediately checks scripts against cached filters!

**Comment by cloudhead** [comment]
> LGTM, I have only a couple of stupid questions.
> 
> ack [6834070](https://github.com/cloudhead/nakamoto/commit/6834070f8f301db23ab0229318c6b60f3ff7cdae)

Sorry actually not done yet!

File: p2p/src/protocol.rs (hunk 2 above)
Comment (vincenzopalazzo) [reply p2p/src/protocol.rs:1110]: Oh yes, the magic is gone, thanks for the clarification

**Comment by vincenzopalazzo** [comment]
>Sorry actually not done yet!

Ops, my fault. Ping me when it is read and you want a review for this PR.

File: p2p/src/protocol/cbfmgr.rs (hunk 1 above)
Comment (cloudhead) [reply p2p/src/protocol/cbfmgr.rs]: This is now done!

**Review by vincenzopalazzo — APPROVED** [review]
ack https://github.com/cloudhead/nakamoto/pull/53/commits/7d8c941a7f9d296e77be79389d5e2b1b65b56fe3

**Pull request merged**